integration-tests/src/test/java/com/datastax/oss/driver/mapper/UdtKeyIT.java

```diff
@@ -0,0 +1,196 @@
+ */
+package com.datastax.oss.driver.mapper;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.api.core.CqlSession;
+import com.datastax.oss.driver.api.core.cql.SimpleStatement;
+import com.datastax.oss.driver.api.mapper.annotations.Dao;
+import com.datastax.oss.driver.api.mapper.annotations.DaoFactory;
+import com.datastax.oss.driver.api.mapper.annotations.DaoKeyspace;
+import com.datastax.oss.driver.api.mapper.annotations.Entity;
+import com.datastax.oss.driver.api.mapper.annotations.Insert;
+import com.datastax.oss.driver.api.mapper.annotations.Mapper;
+import com.datastax.oss.driver.api.mapper.annotations.PartitionKey;
+import com.datastax.oss.driver.api.mapper.annotations.Select;
+import com.datastax.oss.driver.api.testinfra.ccm.CcmRule;
+import com.datastax.oss.driver.api.testinfra.session.SessionRule;
+import com.datastax.oss.driver.categories.ParallelizableTests;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
+import java.util.List;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.rules.RuleChain;
+import org.junit.rules.TestRule;
+
+@Category(ParallelizableTests.class)
+public class UdtKeyIT {
+
+  private static final CcmRule CCM_RULE = CcmRule.getInstance();
+
+  private static final SessionRule<CqlSession> SESSION_RULE = SessionRule.builder(CCM_RULE).build();
+
+  @ClassRule
+  public static final TestRule CHAIN = RuleChain.outerRule(CCM_RULE).around(SESSION_RULE);
+
+  private static RecordDao dao;
+
+  @BeforeClass
+  public static void setup() {
+    CqlSession session = SESSION_RULE.session();
+    for (String ddlQuery :
+        ImmutableList.of(
+            "CREATE TYPE key (value int)",
+            "CREATE TABLE record(key frozen<key> PRIMARY KEY, value int)",
+            "CREATE TABLE multi_key_record(key frozen<list<key>> PRIMARY KEY, value int)")) {
+      session.execute(
+          SimpleStatement.builder(ddlQuery)
+              .setExecutionProfile(SESSION_RULE.slowProfile())
+              .build());
+    }
+
+    TestMapper mapper = new UdtKeyIT_TestMapperBuilder(SESSION_RULE.session()).build();
+    dao = mapper.recordDao(SESSION_RULE.keyspace());
+  }
+
+  @Test
+  public void should_save_and_retrieve_entity_with_udt_pk() {
+    // Given
+    Key key = new Key(1);
+    dao.save(new Record(key, 42));
+
+    // When
+    Record record = dao.findByKey(key);
+
+    // Then
+    assertThat(record.getValue()).isEqualTo(42);
+  }
+
+  @Test
+  public void should_save_and_retrieve_entity_with_udt_collection_pk() {
+    // Given
+    List<Key> key = ImmutableList.of(new Key(1), new Key(2));
+    dao.saveMulti(new MultiKeyRecord(key, 42));
+
+    // When
+    MultiKeyRecord record = dao.findMultiByKey(key);
+
+    // Then
+    assertThat(record.getValue()).isEqualTo(42);
+  }
```

Comment:
I don't see tests that examine the `useNullSavingStrategy`, should we add those?

Reply:
Technically it's not related to this ticket, it would have to be in `NullSavingStrategyIT`. But why not.

Reply:
Actually, `NestedUdtIT` already covers those cases.